inc/Engine/Admin/RocketInsights/Render.php

```diff
@@ -319,4 +341,75 @@ public function truncate_title( string $title ): array {
 	public function render_rocket_insights_column( string $url ): void {
 		echo $this->generate( 'partials/rocket-insights/rocket-insights-column', [ 'url' => $url ] ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
 	}
+
+	/**
+	 * Generates the Rocket Insights column content for post listing pages.
+	 *
+	 * @since 3.20.1
+	 *
+	 * @param string $url The URL of the post.
+	 *
+	 * @return string The rendered HTML for the Rocket Insights column.
+	 */
+	public function get_rocket_insights_column( string $url ): string {
+		// Query for existing row in the database.
+		// Try both with and without trailing slash for compatibility.
+		$url_no_slash   = untrailingslashit( $url );
+		$url_with_slash = trailingslashit( $url );
+
+		$row     = null;
+		$results = $this->query->query( [ 'url' => $url_no_slash ] );
+
+		// If not found without slash, try with slash.
+		if ( empty( $results ) || ! is_array( $results ) ) {
+			$results = $this->query->query( [ 'url' => $url_with_slash ] );
+		}
+
+		if ( ! empty( $results ) && is_array( $results ) ) {
+			$row = $results[0];
+		}
+
+		// Use normalized URL (without trailing slash) for frontend.
+		$normalized_url = $url_no_slash;
+
+		// Get credit availability.
+		$has_credit = $this->plan->has_credit();
```

Comment:
I skipped this in the previous review, but shouldn't this be done at the controller level?

It would also avoid adding the `Query` as a dependency, when it's already the case for the controller

Reply:
Makes sense. I think we have an overall problem of composition, but it's bigger than this PR